flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -954,6 +954,27 @@ public void run() {
 							owner.getName(), checkpointMetaData.getCheckpointId(), asyncDurationMillis);
 				}
 			} catch (Exception e) {
+				// clean up ongoing operator snapshot results and non partitioned state handles
+				for (OperatorSnapshotResult operatorSnapshotResult : snapshotInProgressList) {
+					if (operatorSnapshotResult != null) {
+						try {
+							operatorSnapshotResult.cancel();
+						} catch (Exception cancelException) {
+							e.addSuppressed(cancelException);
+						}
+					}
+				}
+
+				for (StreamStateHandle nonPartitionedStateHandle : nonPartitionedStateHandles) {
```

Comment:
You could use the utility method `StateUtil.bestEffortDiscardAllStateObjects(...)` to reduce code duplication a bit.

I think we should also discard other state objects that completed fuctures could have created, e.g. `StateUtil.bestEffortDiscardAllStateObjects(operatorStatesBackend);` etc. , what do you think?